### DISCUSSION TITLE
Implicit type conversion (reducing dimensions and bool) [MALT NODES]

### DISCUSSION THREAD

**Discussion by Kolupsy** [bnpr/Malt#182] [Q&A] [answered]
I really like how malt now implicitly converts types like blender does but there are still some conversions that are problematic, specifically when reducing the dimensions, for example vec4->float. This comes up the most when using a Sample1Dsample node to manipulate a grayscale gradient. After the color ramp node I always need to add a vec4split. I would suggest using either just the Red channel or the gamma-corrected Value (alpha excluded). The latter option is what blender does in the compositor.
The other conversion that is a little weird is Bool conversion. There does not seem to be a way to convert bool to Int/Float. All connections from bool to those types are being rejected. The only work around i found is to use the inline code node with "float( a )" in the code field.

**Comment by pragma37** [comment, marked as answer]
I have not added implicit conversions when there's not a single unambiguous way to do the conversion.
That said, you can add your own custom conversion functions in your Shader Library.
When 2 sockets of different types are connected, Malt just looks for a function called A_from_B.
So you can just add a function like
```glsl
float float_from_vec4(vec4 v) { return v.x; }
```
or
```glsl
float float_from_vec4(vec4 v) { return luma(v.xyz); }
```
to your Shader Library and the implicit conversion will work.

For bool to int/float you can use the if_else nodes.
But yes, in this case I think they could be added to the built-in lib.

**Reply by Kolupsy** [reply]
I think is is right to be hesitant about these ambiguous conversions (like you did) but I think that because you are consciously connecting two incompatible types, that it would be ok to be met with an 'arbitrary' conversion. I like to think "any conversion is better than no conversion" :D
